### ISSUE TITLE
IE11 throws Exceptions: Automation server can't create object

### ISSUE THREAD

**Issue by tommai78101** [Valve/fingerprintjs2#231]
Browser: Internet Explorer 11 with Edge compatibility

OS Platform: Windows 7

I am getting a couple of errors with the following:

> Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 585, column 13 in https://localhost/Scripts/fingerprint2.js
0x800a01ad - JavaScript runtime error: Automation server can't create object
Exception was thrown at line 1016, column 9 in https://localhost/Scripts/fingerprint2.js
0x800a139e - JavaScript runtime error: a
Exception was thrown at line 1019, column 11 in https://localhost/Scripts/fingerprint2.js
0x800a01b6 - JavaScript runtime error: Object doesn't support property or method 'toSource'
Exception was thrown at line 690, column 9 in https://localhost/Scripts/fingerprint2.js
0x800a139e - JavaScript runtime error: NotSupportedError

My guess is the following:

> Starting with IE11, the navigator object supports plugins and mimeTypes properties. In addition, the window.ActiveXObject property is hidden from the DOM. (This means you can no longer use the property to detect IE11.) IE11 still supports the ActiveXObject object type and you can use it to instantiate ActiveX controls. It is, however, no longer supported as a property of the window object.

Since on line 557, it is trying to find `window.ActiveXObject` property. I think it is that property that is throwing a bunch of exceptions at runtime, but I'm not knowledgeable in this area.

**Comment by Valve** [comment]
thanks, I'll look into it

**Comment by Valve** [comment]
ActiveXObject instantiation is in try/catch block. Please let me know how you're getting those errors, since they should be squelched

**Comment by tommai78101** [comment]
![](http://oi64.tinypic.com/maldec.jpg)

Sorry about the 20 days delay, as I was busy doing other things at work. The image shown above is where I'm hitting the error. The following code is what I've been doing:

```
function test(previousUsagePattern, latestUsagePattern) {
	new Fingerprint2().get(function(hash, components) {
		var dict = {};
		if (latestUsagePattern != -1) {
			dict["previous"] = previousUsagePattern;
			dict["latest"] = latestUsagePattern;
		}

		console.log(hash);
		for (var i = 0; i < components.length; i++) {
			dict[components[i].key] = components[i].value;
			console.log(components[i].key + ": " + components[i].value);
		}

		//Font hash
		var f = "";
		var temp = dict["js_fonts"];
		for (var i = 0; i < temp.length; i++) {
			f += (temp[i]).toString();
			f += ",";
		}
		var fontHash = murmurHash3(f, 31);

		//Canvas hash
		var canvasHash = murmurHash3(dict["canvas"], 31);

		//Plugins hash
		f = "";
		temp = dict["regular_plugins"];
		if (temp == undefined || temp == null)
			temp = dict["ie_plugins"];
		for (var i = 0; i < temp.length; i++) {
			if (temp[i] == null)
				continue;
			f += (temp[i]).toString();
			f += ",";
		}
		var pluginsHash = murmurHash3(f, 31);

		//WebGL hash
		var webglHash = murmurHash3(dict["webgl"], 31);

		//Resolution
		f = "";
		for (var i = 0; i < dict["resolution"].length; i++) {
			f += dict["resolution"][i];
			if (i != dict["resolution"].length - 1)
				f += "x";
		}
		dict["resolution"] = f;

		//Available Resolution
		f = "";
		for (var i = 0; i < dict["available_resolution"].length; i++) {
			f += dict["available_resolution"][i];
			if (i != dict["available_resolution"].length - 1)
				f += "x";
		}
		dict["available_resolution"] = f;

		//Do not track
		if (dict["do_not_track"] == "unknown" || dict["do_not_track"] == "unspecified" || dict["do_not_track"] == undefined || dict["do_not_track"] == null)
			dict["do_not_track"] = "unknown";

		//Indexed database
		if (dict["indexed_db"] == "unknown" || dict["indexed_db"] == "unspecified" || dict["indexed_db"] == undefined || dict["indexed_db"] == null)
			dict["indexed_db"] = -1;

		if (navigator.userAgent.indexOf("MSIE") !== -1 || navigator.appVersion.indexOf("Trident/") > 0) {
			//Browser is using IE11 or under.
			//var f = JSON.stringify(dict["js_fonts"]);
			if (dict["hardware_concurrency"] == "unknown")
				dict["hardware_concurrency"] = -1;

			Services.BackgroundService.HandleDataIE(
				hash, dict["user_agent"], fontHash, canvasHash, pluginsHash, webglHash,
				dict["language"], dict["color_depth"], dict["pixel_ratio"], dict["hardware_concurrency"], dict["resolution"], dict["available_resolution"], dict["timezone_offset"],
				dict["session_storage"], dict["local_storage"], dict["cpu_class"], dict["navigator_platform"], dict["do_not_track"]
			);
			console.log("IE fingerprint logged.");
		} else {
			dict["js_fonts"] = fontHash;
			dict["canvas"] = canvasHash;
			dict["plugins"] = pluginsHash;
			dict["webgl"] = webglHash;
			//Browser is using Google Chrome, Firefox, and others.
			Services.BackgroundService.HandleData(hash, JSON.stringify(dict));
			console.log("Fingerprint logged.");
		}
	});
}

window.onload = test(-1, -1);
```

If you have a simple web page with a Button set with the onclick event to hit test() function, with no parameters, then you should be able to hit the exception.

**Comment by tommai78101** [comment]
I did some more testing. I found out the exception error messages that I'm getting corresponds to all of the try...catch blocks in the fingerprint2.js, and that includes when throwing out the letter "a" at the third exception from the bottom.

I have a hunch that I'm most likely printing the exceptions out to the Console, instead of skipping over them. With a bit of fumbling, I found the culprit. The IE11 debugger was set to break at all exceptions, even when they are all caught and handled correctly. I wasn't aware this was turned on until now.

Disabling this option prevented me from seeing the exception messages, and everything went running just fine. I'm closing this issue, as it's just a false alarm and not as serious as I thought it was. Sorry for your time.
